src/bytecode.rs

```diff
@@ -633,11 +634,29 @@ fn read_opcodes<'a>(
                 };
                 Opcode::Newarray(primitive_type)
             }
-            0xbd => Opcode::Anewarray(read_cp_classinfo(code, &mut ix, pool)?),
+            0xbd => Opcode::Anewarray({
+                let ty = read_cp_classinfo(code, &mut ix, pool)?;
+                match FieldType::parse(&ty) {
+                    Ok(ty) => ty,
+                    Err(_) => FieldType::Ty(Ty::Object(ty)),
```

Comment:
This doesn't look right to me. Conceptually the classinfo field contains the name of a class, which is different from a field type. For instance if the class happened to be called "B", parsing it as a field type would turn it into `FieldType::Ty(Ty::Base(BaseType::Byte))` which is incorrect. On top of that this approach of handling `Err` and assuming it is an `Ty::Object` seems not very robust. `Err` can be returned for any number of reasons and is not a good way to handle control flow. You're basically hacking around the fact that ClassInfo structures contain a different syntax (e.g. `java/lang/String`) than what field type is intended to parse (`Ljava/lang/String;`) but this is not a good solution.

Reply:
Having to manually parse the Cow of anewarray, instanceof and checkcast also seems a little strange though

Reply:
Yeah like I said in the comment below, I'm not opposed to putting in a structured type here instead of the Cow. I'm just saying using `FieldType` is wrong, and it should be a different structured type that precisely represents the possible range of values here.

Reply:
How is it wrong? can you not checkcast from Object to [J or [B?

Reply:
> For instance if the class happened to be called "B", parsing it as a field type would turn it into `FieldType::Ty(Ty::Base(BaseType::Byte))` which is incorrect.

Reply:
Yes, It would.
Write something that works then.
As far as I am concerned there is overlap in the possible values. You have to decide, if B is a BasicType or the class B.

I chose to do it this way, since it was easiest to implement. I didn't have to write a whole new function, but only modify a couple of lines.

I still believe that something that those opcodes should parse the array information.

A viable way to do this might be to just

```rust
let mut ty = ty;
let mut dims = 0;
while let Some(new_ty) = ty.split_prefix("[") {
  dims+=1;
  ty=new_ty;
}
```

Reply:
> I chose to do it this way, since it was easiest to implement. I didn't have to write a whole new function, but only modify a couple of lines.

You're welcome to do what you want, in your own fork. If you want to merge your changes to my implementation you'll have to write code that conforms to my standards. This is how open source works.

> I still believe that something that those opcodes should parse the array information.

Like I said I'm not at all opposed to this. I just want the implementation to use types with precision.

> A viable way to do this might be to just
> 
> ```rust
> let ty = ty.split_prefix("L").unwrap_or(ty);
> let ty = ty.split_suffix(";").unwrap_or(ty);
> let mut ty = ty;
> let mut dims = 0;
> while let Some(new_ty) = ty.split_prefix("[") {
>   dims+=1;
>   ty=new_ty;
> }
> ```

There's a number of problems with this code, starting with the fact that the class name in these instructions is not encoded with `L...;`. If you get something that works feel free to open a new PR and I can take a look.

Reply:
The code works when you use `stip_prefix` instead of `split_prefix`. I added the `L...;` bit as an extra bit of precaution, just to be sure.

```rust
fn main() {
    let ty = "[[[test";
    let mut ty = ty;
    let mut dims = 0;
    while let Some(new_ty) = ty.strip_prefix("[") {
        dims+=1;
        ty=new_ty;
    }
    
    let ty = ty.strip_prefix("L").unwrap_or(ty);
    let ty = ty.strip_suffix(";").unwrap_or(ty);
    
    println!("{ty}, {dims}");
}
```

Reply:
> Like I said I'm not at all opposed to this. I just want the implementation to use types with precision.

I proposed multiple solutions now, but you find none sufficient.
Tbh I do not know what you want from me at this point.

Reply:
Also are you saying that the following sample opcode is wrong too? (also taken from #41)

`21: checkcast     #21                 // class "[Ljava/lang/String;"`

Reply:
![image](https://github.com/user-attachments/assets/94a56398-7d74-4abc-a31a-43c620d710a6)

Reply:
> The code works when you use `strip_prefix` instead of `split_prefix`. I added the `L...;` bit as an extra bit of precaution, just to be sure.
> 
> ```rust
> fn main() {
>     let ty = "[[[test";
>     let mut ty = ty;
>     let mut dims = 0;
>     while let Some(new_ty) = ty.strip_prefix("[") {
>         dims+=1;
>         ty=new_ty;
>     }
>     
>     let ty = ty.strip_prefix("L").unwrap_or(ty);
>     let ty = ty.strip_suffix(";").unwrap_or(ty);
>     
>     println!("{ty}, {dims}");
> }
> ```


Try this java code:

```
class L {
    public L tryCast(Object o) {
        L l = (L)o;
        return l;
    }
}
```

Look at what the checkcast instruction generates, the string value that you get, and what your parsing code does to it.

Reply:
With my new parsing code (ReferenceType) I'd expect it to correctly resolve to the Object Variant of ReferenceType.

I did miss a couple of cases of FieldType though, that should be replaced with ReferenceType.